### PR TITLE
Feature/xinnanli/account name format

### LINKED ISSUE #218: The Ability to Customize the Account Naming Setting
Hub thread: https://powerofus.force.com/0D58000002atMIl

Description from the existing item in the internal backlog that has now been closed as duplicate: "Allow users to, for example, add the Contact's student ID to the name of the Administrative Account."

### PR DESCRIPTION
**Pull request by weizsforg**
# Critical Changes
# Changes
- We now support customizing the format of Administrative Account and Household Account names. In HEDA Settings, on the System tab, choose a preset naming option or create your own custom naming convention. For details, see [Customize Administrative and Household Account Names](https://powerofus.force.com/articles/Resource/HEDA-Customize-Administrative-and-Household-Account-Names). Thank you, Jennifer Sanchez (Attain), for promoting this idea in the Power of Us Hub! https://powerofus.force.com/hub-ideas/1158

   If you don't take the actions described below, Administrative Account names continue to be automatically updated whenever the Contact's last name changes. And Household Account names are not automatically updated unless you enable the new setting, **Automatic Household Naming**, which renames the Household Account whenever a new Contact is added or an existing Contact is renamed, deleted, or moved to another Account.

    **Action required:**
    For the Administrative Account Name Format field, select the default value that you'd like to use. Changing the default name format does not retroactively update existing Administrative Account names.

    For the Household Account Name Format field, select the default value that you'd like to use. Decide whether to enable Automatic Household Naming. If you leave the setting disabled, when a Household Account is created, we continue to _name_ it using the default name format you've specified but we do not _rename_ the Account when Contacts within the Household change. Whether automatic naming is enabled or disabled, changing the default name format does not retroactively update existing Household Account names. Note: We don't currently support the use of custom fields in custom name formats for Household Accounts.

# Issues Closed
Fixes #218
# New Metadata
CMP_Acc_Naming_Dropdown
Hierarchy_Settings__c.Admin_Account_Naming_Format__c
Hierarchy_Settings__c.Admin_Other_Name_Setting__c
Hierarchy_Settings__c.Household_Account_Naming_Format__c
Hierarchy_Settings__c.Household_Other_Name_Setting__c
None
acctNamingOther
adminAccNameFormat
firstNameLastNameAdminACC
firstNameLastNameFamily
firstNameLastNameHH
hhAccNameFormat
hhFamily
lastNameAdminAcc
lastNameFamily
lastNameFirstNameAdminAcc
lastNameFirstNameFamily
lastNameFirstNameHH
lastNameHH
# Deleted Metadata
# Testing Notes
1. Deploy the HEAD package to Dev org or Scratch org
2. Go to HEDA setting page and switch to system tab
3. Select Admin account or Household account as default record type
4. Select naming setting for admin account and household account
4. Create a contact and make sure the name of account is correct

### CODE REVIEW

File: src/classes/ACCT_IndividualAccounts_TDTM.cls [hunk 1]
```diff
@@ -142,6 +153,14 @@ public class ACCT_IndividualAccounts_TDTM extends TDTM_Runnable {
                     if (c.AccountId != oldContacts[i].AccountId) {
                         if (c.AccountId != null) {
                             listAccountIdHHToUpdate.add(c.AccountId);
+                            if (c.Account.RecordTypeId == UTIL_Describe.getHhAccRecTypeID()
+                                    && UTIL_CustomSettingsFacade.getSettings().Automatic_Household_Naming__c != false){
+                                setAccountIdHHToRename.add(c.AccountId);
+                            }
```
Comment (weizsforg) [inline src/classes/ACCT_IndividualAccounts_TDTM.cls]: I think in here, we also need to update oldContacts[i].Account name if oldContacts[i].AccountId != null.

File: src/aura/STG_CMP_System/STG_CMP_System.cmp [hunk 2]
```diff
@@ -270,5 +275,94 @@
...
+        <div class="slds-col slds-size--1-of-2">
+            <aura:if isTrue="{!v.namespacePrefix == 'hed__'}">
+                <ui:outputText value="{!$Label.hed.adminAccNameFormat}"/>
+            <aura:set attribute="else">
+                <ui:outputText value="{!$Label.c.adminAccNameFormat}"/>
+            </aura:set>
+            </aura:if>
+        </div>
+        
+        <div class="slds-col slds-size--1-of-2" >
```
Comment (weizsforg) [inline src/aura/STG_CMP_System/STG_CMP_System.cmp]: I think there is a small UI issue here. System does not show the current setting. 

Steps to reproduce:
1. Deploy Master branch to the org
2. Deploy the Account Naming branch to the org
3. Go to Naming Setting, click Edit and click Save
4. The Naming settings in the UI shows blank. But the value actually saved in the custom settings

**Comment by weizsforg** [comment]
@xinnanlisf Please have a look about the comments in this PR. Also, looks like this branch is out of date. Can you update this branch?

**Comment by weizsforg** [comment]
@xinnanlisf Maybe we should ask someone to review the 'Changes' in this PR?

File: src/classes/ACCT_IndividualAccounts_TDTM.cls (hunk 1 above)
Comment (xinnanlisf) [reply src/classes/ACCT_IndividualAccounts_TDTM.cls]: Good catch!

File: src/aura/STG_CMP_System/STG_CMP_System.cmp (hunk 2 above)
Comment (xinnanlisf) [reply src/aura/STG_CMP_System/STG_CMP_System.cmp]: Sorry which setting are you referring to? I can't seem to reproduce it.

**Comment by xinnanlisf** [comment]
@lianeyee 

We might need to update the release notes. Household accounts will now be updated if a child contact is deleted or moved to another account(This account will also be updated!).

**Comment by lianeyee** [comment]
We definitely need to clean up the release notes for public consumption. It's on my radar for today. Thanks, @xinnanlisf

**Comment by lianeyee** [comment]
@xinnanlisf @weizsforg When you have a chance, please review the revised release notes. I'm afraid they got a bit long and involved. But I think it's better to be as clear and specific as possible about how Household Account naming behaves.

**Comment by lianeyee** [comment]
@xinnanlisf It just occurred to me that the help text for Automatic Household Naming might be  incorrect now. 

According to the WI, it said: "Enables automatic naming of Household Accounts, based on the names of Contacts in the Household. If enabled, we also automatically rename Household Accounts when new Contacts are added or existing Contacts are renamed."

According to the behavior now, shouldn't it say something like this? "Enables automatic renaming of Household Accounts, based on the names of Contacts in the Household. If enabled, we automatically rename Household Accounts when a new Contact is added to the Household or an existing Contact is renamed, deleted, or moved to another Household."

**Comment by xinnanlisf** [comment]
@lianeyee 

Good catch. I will go ahead and make the changes. :)

**Comment by weizsforg** [comment]
@xinnanlisf @lianeyee I think there are still some questions when the Contact is renamed, deleted, or moved to another Account for both Admin and HH Account model.  

Currently:
1. For Admin Account Model: We only update Account Name when LastName of the Contact changes. We do not update Account Name when Contact is renamed, deleted, or moved to another Account.

2. For HH Account Model: We included the logic to handle Contact is renamed, deleted, or moved to another Account in this PR. But I think there is an issue when we move the Contact from One Account to Another Account.

It will be good to document the expect behavior (maybe with an example) when Contact is renamed, deleted, or moved to another Account for both Admin and HH Account model in HEDA. 

Any ideas?

**Comment by xinnanlisf** [comment]
@weizsforg 

Are you questioning about the help text? I think it is used only for Household account, correct?

**Comment by weizsforg** [comment]
@xinnanlisf Actually, it is about the actual logic. 

In the "Changes" of this PR and the Quip doc, we mention "Administrative Account names continue to be automatically updated whenever a new Contact is added to an Account or an existing Contact is renamed, deleted, or moved to another Account. ". 

But I don't think we currently doing that. We only update Admin Acc Name when Contact LastName changes.

**Comment by lianeyee** [comment]
That's my error then. For Admin Account, it sounds like it should be "Administrative Account names continue to be automatically updated whenever the Contact's name changes." (If they choose a name format that includes First Name, we shouldn't specify just Last Name, right?)

**Comment by weizsforg** [comment]
I think the code logic only handle the Contact LastName changes of the Admin Account model.  What do you think? @xinnanlisf Should we just mention Name instead of LastName?

**Comment by weizsforg** [comment]
Thanks for all the great work! @xinnanlisf
Looks like I can not be the reviewer since this PR is created by me. Can you approve it @xinnanlisf ? 

@annejackson This PR is ready for QA now. I think one important thing to test is to make sure the default behavior does not change when user do nothing after we ship this feature.

**Comment by annejackson** [comment]
Hey @xinnanlisf - I'm not ready to test (or pass) this yet, but I just noticed that the "merge pull request" button is disabled...do we know why?

**Comment by xinnanlisf** [comment]
@annejackson 

That's mostly because @weizsforg reverted the pull request and someone else need to approve it. I just approved it and it should be ready for merge.

**Comment by annejackson** [comment]
Yep, I see that it's available to merge now. Thanks @xinnanlisf

**Comment by annejackson** [comment]
@xinnanlisf saw this issue while using the Other feature. See screenshots. 
settings when saw bug
bug when adding contact with other settings

**Comment by xinnanlisf** [comment]
@annejackson 

Should be good now. Good catch! I would suggest let @weizsforg do another code review before merging it. :)

**Comment by annejackson** [comment]
OK @xinnanlisf I can move into my other test cases that I have left using my current scratch org.  @ceiroa  since Wei is out on PTO, can you or someone else do a review, or how does this normally work?

**Comment by ceiroa** [comment]
@annejackson @xinnanlisf Yes, I'll take a look.

**Comment by annejackson** [comment]
@xinnanlisf  here's the info, screenshots I told you about earlier. 
settings for bug2
bug2
fyi @ceiroa

**Comment by xinnanlisf** [comment]
@annejackson  should be good now! Thanks for your help.

**Comment by annejackson** [comment]
@ceiroa  can you review the changes? Thanks!

**Comment by ceiroa** [comment]
@annejackson The changes have been reviewed, and this is all ready for one last review next Monday. Thanks!

**Comment by xinnanlisf** [comment]
@lianeyee 
I just updated the release notes. Could you do a quick review please? Also, regarding the document you have been working on, we might need to add a note in it as well. Let me know if you need any help or any questions.

**Comment by lianeyee** [comment]
Thanks for updating the release note, @xinnanlisf. I moved that item down to the Household Account section because this limitation applies only to Household Accounts. I will add a similar note to the documentation.

**Comment by annejackson** [comment]
Testing has passed in the scratch org. Merging this pull request. Nice work @xinnanlisf. 
Fyi @ceiroa

**Pull request merged**